Functionality to manually add URL's to the feeds overview

FokkoVeegens opened the pull request:
This pull request introduces a new GitHub Actions workflow to automate the addition of URLs to a `data.json` file and updates the front-end to handle and display these URLs. The most important changes include the creation of the workflow file, modifications to the `data.json` structure, and updates to the front-end logic to merge and display manual URLs.

### GitHub Actions Workflow:
* [`.github/workflows/add-page-to-feed.yml`](diffhunk://#diff-01ed6b2c057e6796b5d10ef7ddcfc48fa747cc637ace1dedd79b6df1a1398813R1-R99): Added a new workflow that triggers on issue creation with the label 'add-url', extracts and validates the URL from the issue body, fetches the URL title, updates `data.json`, and commits the changes. It also comments on the issue and closes it based on the success or failure of the operation.

### Data Structure Update:
* [`public/data.json`](diffhunk://#diff-e37e3a9264ad80f2d0c18ae3e57737ba5ad9c2a752cd3e93add34ed638f23c04R316): Added a new `manualUrls` array to store manually added URLs.

### Front-End Updates:
* [`src/pages/Index.jsx`](diffhunk://#diff-d43d160cd064cf3eade3358320e2f4df156846513ccdf09fc890d704fb7ce5f4R20): Updated the component to merge `manualUrls` with existing RSS feed items, ensuring unique display and handling of these URLs. This includes new functions for merging URLs, updating state, and rendering the combined list of items. [[1]](diffhunk://#diff-d43d160cd064cf3eade3358320e2f4df156846513ccdf09fc890d704fb7ce5f4R20) [[2]](diffhunk://#diff-d43d160cd064cf3eade3358320e2f4df156846513ccdf09fc890d704fb7ce5f4L28-R57) [[3]](diffhunk://#diff-d43d160cd064cf3eade3358320e2f4df156846513ccdf09fc890d704fb7ce5f4R95-R108) [[4]](diffhunk://#diff-d43d160cd064cf3eade3358320e2f4df156846513ccdf09fc890d704fb7ce5f4L81-R120) [[5]](diffhunk://#diff-d43d160cd064cf3eade3358320e2f4df156846513ccdf09fc890d704fb7ce5f4R142-R154) [[6]](diffhunk://#diff-d43d160cd064cf3eade3358320e2f4df156846513ccdf09fc890d704fb7ce5f4L161-L176)

These changes collectively enhance the functionality by automating URL additions and integrating them seamlessly into the existing news feed system.

rajbos commented on src/pages/Index.jsx:
```diff
@@ -148,30 +187,36 @@ const Index = () => {
-              {feed.name} ({feedCounts[feed.name] || 0})
-            </button>
-          ))}
+            feeds.filter(feed => !feed.cors)
```
I thought we removed that filter in a previous update?

FokkoVeegens replied:
Yes, I will update this branch with the changes in main ;-)

rajbos commented:
Checking the intent here: do you want to add a single url as a feed item into the `data.json` and then display it (and then where does it show up?), or do you want to add a complete RSS feed into the `data.json` and display all items in that feed? 

Separate comment: this is rather magic behavior as it is right now, so I would like to add an issue template for this use case, that already fills in the label for the user and then does the rest. But that would be a security issue for people creating issues without write access 🤔. Should be less of an issue then for ourselves. 

Perhaps this behavior should be documented then in the issue body of the template, and then the label is something the repo admins can add after reviewing the change. That way we prevent random users of creating issues that end up in our repo 😃.

FokkoVeegens commented:
Fixed the issues mentioned in the previous comment. Now when people with write/maintain/admin permissions add a URL, it's applied directly in the default branch (currently a hardcoded branch, needs to be changed). If somebody with read permissions adds a URL, a PR in a separate branch is created.

### GitHub Actions Workflow

* [`.github/ISSUE_TEMPLATE/add-url.yml`](diffhunk://#diff-00348d44c83254134af3b5941e73ec27ba5309477c21e102fdb4877fb3738e92R1-R13): Added a new issue template for adding URLs to the newsfeed.
* [`.github/ISSUE_TEMPLATE/config.yml`](diffhunk://#diff-1c0d972ee49103af56fd608a77a28e1eb12f6908f263f0f183d46868fdcd8ea2R1-R5): Updated the configuration to include a contact link for the new "Add URL" issue template.

### Documentation Update

* [`README.md`](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R18-R21): Added a section explaining how to add URLs to the newsfeed using the new issue form.

rajbos commented on .github/workflows/add-page-to-feed.yml:
```diff
@@ -0,0 +1,142 @@
+          ERROR_MESSAGE="${{ env.ERROR_MESSAGE }}"
+          ISSUE_NUMBER=${{ github.event.issue.number }}
+          COMMENT="The addition of the URL is aborted due to the following error: $ERROR_MESSAGE"
+          gh issue close $ISSUE_NUMBER --comment "$COMMENT"
```
Why close the issue? Adding the comment can help understand why it fails, which gives the requestor the option to update the body and relabel the issue to retrigger the workflow.

FokkoVeegens replied:
We can do that of course, that would need some refactoring, as the workflow is now only triggered for new issues. I'll add this when I have time again ;-)

rajbos commented on README.md:
```diff
@@ -15,3 +15,7 @@ To run the application in a GitHub Codespace, simply open the repository in a Co
+
+## Adding URLs to the newsfeed
+
+It is possible to add a single URL to the newsfeed by using the issue form. Simply click on the "Issues" tab and create a new issue with the URL you would like to add. The issue will be automatically closed and the URL will be added to the newsfeed.
```
```suggestion
It is possible to add a single URL to the newsfeed by using the issue form. Simply click on the "Issues" tab and create a new issue with the URL you would like to add. The issue will be automatically closed and the URL will be added to the newsfeed, if you have write rights on this repository.
```
